splash/qtutils.py

```diff
@@ -299,3 +321,47 @@ def qt_header_items(request_or_reply):
         (name, request_or_reply.rawHeader(name))
         for name in request_or_reply.rawHeaderList()
     ]
+
+
+def qt_send_key(key, target):
+    """
+    Send a key event that might be defined using emacs keyboard macro syntax
+    to target qt object. For example:
+        - RET -> Qt.Key_Return
+        - SPC -> Qt.Key_Space
+        - <FooBar> -> Qt.Key_FooBar
+    See: http://doc.qt.io/qt-5/qt.html#Key-enum for valid key types
+    """
+    # Try to match EDMACRO shortcut keys
+    key_type = EDMACRO_SHORTCUTS.get(key)
+    if key_type:
+        text = QT_KEY_INPUTS.get(key_type, '')
+        return qt_send_text(text, target, key_type)
+
+    # Try to match "<Key_Name>"
+    key_match = re.match(r'^\<(.+)\>$', key)
+    if key_match:
+        key_type = getattr(Qt, 'Key_%s' % key_match.group(1),
+                           Qt.Key_unknown)
+        text = QT_KEY_INPUTS.get(key_type, '')
+        return qt_send_text(text, target, key_type)
+
+    # All above failed, send as fallback input
+    return qt_send_text(key, target)
+
+
+def qt_send_text(text, target, key_type=0):
+    """
+    Send text as key input event to target qt object, as if generated by
+    `key_type`. Key type defaults to 0, meaning "the event is not a result of a
+    known key; for example, it may be the result of a compose sequence or
+    keyboard macro."
+    """
+    modifiers = QApplication.keyboardModifiers()
+    text = list(text) or ['']
+    for x in text:
+        event = QKeyEvent(QEvent.KeyPress, key_type, modifiers, x)
+        QApplication.postEvent(target, event)
+        # Key release does not generate any input
+        event = QKeyEvent(QEvent.KeyRelease, key_type, modifiers, '')
+        QApplication.postEvent(target, event)
```

Comment:
@kmike: Turns out, regardless of the key event sent, browser understands a keydown + keyup as a correct keypress. Assuming that and being the implementation straightforward, it makes sense to send an event per key, so I added it.